Sources/SemanticIndex/SemanticIndexManager.swift

```diff
@@ -185,31 +185,41 @@ public final actor SemanticIndexManager {
   }
 
   public func filesDidChange(_ events: [FileEvent]) async {
-    let filesToReIndex = await filesToReIndex(changedFiles: events.map(\.uri))
-
+    // We only re-index the files that were changed and don't re-index any of their dependencies. See the
+    // `Documentation/Files_To_Reindex.md` file.
+    let changedFiles = events.map(\.uri)
     // Reset the index status for these files so they get re-indexed by `index(files:priority:)`
-    for uri in filesToReIndex {
+    for uri in changedFiles {
       indexStatus[uri] = nil
     }
-    await scheduleBackgroundIndex(files: filesToReIndex)
+    await scheduleBackgroundIndex(files: changedFiles)
   }
 
-  /// Returns the files that should be re-indexed if the given files have been modified.
+  /// Returns the files that should be indexed to get up-to-date index information for the given files.
   ///
-  /// - SeeAlso: The `Documentation/Files_To_Reindex.md` file.
-  private func filesToReIndex(changedFiles: [DocumentURI]) async -> Set<DocumentURI> {
+  /// If `files` contains a header file, this will return a `FileToIndex` that re-indexes a main file which includes the
+  /// header file to update the header file's index.
+  private func filesToIndex(toCover files: some Collection<DocumentURI>) async -> [FileToIndex] {
     let sourceFiles = Set(await buildSystemManager.sourceFiles().map(\.uri))
-    let filesToReIndex = await changedFiles.asyncMap { (uri) -> [DocumentURI] in
+    let filesToReIndex = await files.asyncCompactMap { (uri) -> FileToIndex? in
       if sourceFiles.contains(uri) {
-        // If this is a source file, re-index it
-        return [uri]
+        // If this is a source file, just index it.
+        return FileToIndex(uri: uri, mainFile: nil)
+      }
+      // Otherwise, see if it is a header file. If so, index a main file that that imports it to update header file's
+      // index.
+      // Deterministically pick a main file. This ensures that we always pick the same main file for a header. This way,
+      // if we request the same header to be indexed twice, we'll pick the same unit file the second time around,
+      // realize that its timestamp is later than the modification date of the header and we don't need to re-index.
+      let mainFile = index.checked(for: .deletedFiles)
+        .mainFilesContainingFile(uri: uri, crossLanguage: false)
+        .sorted(by: { $0.stringValue < $1.stringValue }).first
+      guard let mainFile else {
+        return nil
       }
-      // Otherwise, see if it is a header file. If so, re-index all the files that import it.
-      // We don't want to re-index `.swift` files that depend on a header file, similar to how we don't re-index a Swift
-      // module if one of its dependent modules has changed.
-      return index.checked(for: .deletedFiles).mainFilesContainingFile(uri: uri, crossLanguage: false)
-    }.flatMap { $0 }
-    return Set(filesToReIndex)
```

Comment:
Given we're no longer uniquing, is there any concern that if e.g 5 headers get modified, and those headers are all included by a single main file, that we could end up running 5 concurrent indexing jobs for that main file?

Reply:
I also didn't really think about the scheduling logic here :) Looking at it again, we'll block the requests on each other, and then they should all be considered up-to-date

Reply:
We don’t actually, because they all have different `FileToIndex.uri` and thus they would actually be executing in parallel. Fixing that now.